src/high-level/tensor.lisp

```diff
@@ -0,0 +1,130 @@
+                             (rank shape size order storage))
+                         (:copier ,copy-sym))
+         (storage nil :type (tensor-storage ,type)))
+       #+sbcl (declaim (sb-ext:freeze-type ,name))
+       
+       (defmethod storage ((m ,name))
+         (,storage-sym m))
+
+       (defmethod element-type ((m ,name))
+         (declare (ignore m))
+         ',type)
+       
+       (defmethod make-tensor ((class (eql ',name)) shape &key initial-element order storage)
+         (policy-cond:policy-if
+          (< speed safety)
+          (check-type shape shape)
+          nil)
+         (let ((size (reduce #'* shape)))
+           (funcall #',constructor-sym
+                    (length shape)
+                    shape
+                    size
+                    (or order :column-major)
+                    (or
+                     storage
+                     (apply #'make-array
+                            size
+                            :element-type ',type
+                            (if initial-element
+                                (list :initial-element (coerce initial-element ',type))
+                                nil))))))
+       (defmethod cast ((tensor ,name) (class (eql ',name)))
+         (declare (ignore class))
+         tensor)
+
+       ;; TODO: This does not allow for args. Make this allow for args.
+       (defmethod copy-tensor ((m ,name) &rest args)
+         (declare (ignore args))
+         (let ((new-m (,copy-sym m)))
+           (setf (,storage-sym new-m)
+                 (make-array (tensor-size m) :element-type (element-type m)))
+           new-m))
+
+       (defmethod deep-copy-tensor ((m ,name) &rest args)
+         (declare (ignore args))
+         (let ((new-m (,copy-sym m)))
+           (setf (,storage-sym new-m)
+                 (alexandria:copy-array (,storage-sym m)))
+           new-m))
+
+       (defmethod tref ((tensor ,name) &rest pos)
+         (let ((index (case (tensor-order tensor)
+                        (:row-major (row-major-index pos (tensor-shape tensor)))
+                        (:column-major (column-major-index pos (tensor-shape tensor))))))
+           (aref (,storage-sym tensor) index)))
+
+       (defmethod (setf tref) (new-value (tensor ,name) &rest pos)
+         (let ((index (case (tensor-order tensor)
+                        (:row-major (row-major-index pos (tensor-shape tensor)))
+                        (:column-major (column-major-index pos (tensor-shape tensor))))))
+           (setf (aref (,storage-sym tensor) index)
+                 new-value))))))
+
+(defun pprint-tensor (stream tensor &optional colon-p at-sign-p)
+  "Pretty-print a matrix MATRIX to the stream STREAM."
+  (declare (ignore colon-p)
+           (ignore at-sign-p))
+  (pprint-logical-block (stream nil)
+    (print-unreadable-object (tensor stream :type t)
+      (format stream "(~{~D~^x~})" (shape tensor)))))
+
+(set-pprint-dispatch 'tensor 'pprint-tensor)
```

Comment:
are you a wizard?

Reply:
It was clear that he was from the very first day! :-P